### PR TITLE
[Refactor]: Unified parameter initialization

### PR DESCRIPTION
**Pull request by hhaAndroid**
## Motivation
Refactoring model initialization methods based MMCV. For more information, please refer to https://github.com/open-mmlab/mmcv/pull/780

## Dependency

- [x] https://github.com/open-mmlab/mmcv/pull/780
- [x]  https://github.com/open-mmlab/mmcv/pull/884
- [x] https://github.com/open-mmlab/mmcv/pull/893
- [x]  https://github.com/open-mmlab/mmcv/pull/899
- [x]  https://github.com/open-mmlab/mmcv/pull/902
- [x] https://github.com/open-mmlab/mmcv/pull/918
- [x] https://github.com/open-mmlab/mmcv/pull/930
- [x] https://github.com/open-mmlab/mmcv/pull/959
- [x] https://github.com/open-mmlab/mmcv/pull/971

## Models

Models that have been refactored as follow:

- [x] ATSS
- [x] CARAFE 
- [x] Cascade R-CNN
- [x] Faster R-CNN
- [x] DCN
- [x] Double Head
- [x] Dynamic R-CNN
- [x] Empirical Attention
- [x] Foveabox
- [x] FreeAnchor
- [x]  FASF
- [x] GCNet
- [x] GFL
- [x] GN
- [x] GROIE 
- [x] Guided Anchoring
- [x] HRNet
- [x] HTC
- [x] MS R-CNN
- [x] NAS FPN
- [x] PAA
- [x] PAFPN
- [x] PISA
- [x] RepPoints
- [x] ResneSt
- [x] RetinaNet
- [x] GHM
- [x] RPN
- [x] SABL
- [x] SCNet
- [x] TridentNet
- [x] Sparse R-CNN
- [x] VFNet
- [x] YOLOV3
- [x] FPG
- [x] cascade_rpn
- [x] centripetalnet
- [x] cornernet
- [x] detectors
- [x] detr
- [x] nas_fcos
- [x] gn+ws (40.2/36.2) vs (40.6/36.6)
- [x] grid_rcnn
- [x] libra_rcnn
- [x] point_rend
- [x] regnet
- [x] res2net
- [x] ssd
- [x] yolact

All models have been checked and passed！

### CODE REVIEW

**Comment by ZwwWayne** [comment]
1. Fix lint
2. Involve @MeowZheng to have a look.

**Comment by hhaAndroid** [comment]
@MeowZheng @jshilong  Please review the one-stage algorithm first, I tagged todo where I couldn't be sure.

File: mmdet/models/backbones/detectors_resnet.py [hunk 1]
```diff
@@ -56,12 +57,8 @@ def __init__(self,
...
                 stride=1,
                 bias=True)
-        self.init_weights()
-
-    def init_weights(self):
-        """Initialize the weights."""
-        if self.rfp_inplanes:
-            constant_init(self.rfp_conv, 0)
+            self.init_cfg = dict(
+                override=dict(type='Constant', val=0, name='rfp_conv'))
```
Comment (MeowZheng) [inline mmdet/models/backbones/detectors_resnet.py]: in __init__ 
self.init_cfg = dict(type='Constant', val=0), override=dict(name='rfp_conv'))
if inherit `BaseModule`, there is no need to call init_weights(),

File: mmdet/models/backbones/detectors_resnet.py [hunk 2]
```diff
@@ -230,13 +227,11 @@ def __init__(self,
                  stage_with_sac=(False, False, False, False),
                  rfp_inplanes=None,
                  output_img=False,
-                 pretrained=None,
```
Comment (MeowZheng) [inline mmdet/models/backbones/detectors_resnet.py:233]: for backward compatibility, please do not delete kwarg pretrained. 
in __init__add 
```
        if pretrained is not None:
            warnings.warn('DeprecationWarning: pretrained is a deprecated \
                key, please consider using init_cfg')
            self.init_cfg = dict(type='Pretrained', checkpoint=pretrained)

File: mmdet/models/backbones/res2net.py [hunk 3]
```diff
@@ -305,47 +301,41 @@ def __init__(self,
...
+                    dict(type='Kaiming', layer='Conv2d'),
+                    dict(
+                        type='Constant',
+                        val=1,
+                        layer=['_BatchNorm', 'GroupNorm'])
+                ]
+            elif not isinstance(init_cfg, list):
+                self.init_cfg = [init_cfg]
+
+            # TODO: dcn does not support init_cfg mode
```
Comment (MeowZheng) [inline mmdet/models/backbones/res2net.py:329]: please clearify.
`# TODO: dcn conv_offset cannot initialize with init_cfg`

File: mmdet/models/backbones/trident_resnet.py [hunk 4]
```diff
@@ -33,8 +34,10 @@ def __init__(self,
                  stride=1,
                  trident_dilations=(1, 2, 3),
                  test_branch_idx=1,
-                 bias=False):
-        super(TridentConv, self).__init__()
+                 bias=False,
+                 init_cfg=dict(
+                     type='Kaiming', distribution='uniform', mode='fan_in')):
```
Comment (MeowZheng) [inline mmdet/models/backbones/trident_resnet.py:39]: add layer key

File: mmdet/models/dense_heads/nasfcos_head.py [hunk 5]
```diff
@@ -17,6 +16,38 @@ class NASFCOSHead(FCOSHead):
...
+                             name='cls_convs'),
+                         dict(
+                             type='Kaiming',
+                             a=1,
+                             mode='fan_in',
+                             nonlinearity='leaky_relu',
+                             distribution='uniform',
+                             layer=['ConvModule', 'Conv2d'],
+                             name='reg_convs')
+                     ]),
```
Comment (MeowZheng) [inline mmdet/models/dense_heads/nasfcos_head.py:47]: little unelegant.
may do like
```
init_cfg = None):
if init_cfg is not None:
    init_cfg = .....
super(NASFCOSHead, self).__init__(*args, init_cfg=init_cfg, **kwargs)

File: mmdet/models/dense_heads/reppoints_head.py [hunk 6]
```diff
@@ -51,6 +51,15 @@ def __init__(self,
...
                  moment_mul=0.01,
+                 init_cfg=dict(
+                     type='Normal',
+                     layer='Conv2d',
+                     std=0.01,
+                     override=dict(
+                         type='Normal',
+                         name='reppoints_cls_out',
+                         std=0.01,
+                         bias_prob=0.01)),
```
Comment (MeowZheng) [inline mmdet/models/dense_heads/reppoints_head.py:62]: for long init_cfg do like last comment

File: mmdet/models/dense_heads/vfnet_head.py [hunk 7]
```diff
@@ -96,6 +96,15 @@ def __init__(self,
...
                      strides=[8, 16, 32, 64, 128]),
+                 init_cfg=dict(
+                     type='Normal',
+                     layer='Conv2d',
+                     std=0.01,
+                     override=dict(
+                         type='Normal',
+                         name='vfnet_cls',
+                         std=0.01,
+                         bias_prob=0.01)),
```
Comment (MeowZheng) [inline mmdet/models/dense_heads/vfnet_head.py:107]: long init_cfg

File: mmdet/models/dense_heads/yolact_head.py [hunk 8]
```diff
@@ -59,6 +59,7 @@ def __init__(self,
                  use_ohem=True,
                  conv_cfg=None,
                  norm_cfg=None,
+                 init_cfg=dict(type='Xavier', distribution='uniform', bias=0),
```
Comment (MeowZheng) [inline mmdet/models/dense_heads/yolact_head.py:62]: add layer

File: mmdet/models/dense_heads/yolact_head.py [hunk 9]
```diff
@@ -477,8 +471,9 @@ def __init__(self,
                  loss_segm=dict(
                      type='CrossEntropyLoss',
                      use_sigmoid=True,
-                     loss_weight=1.0)):
-        super(YOLACTSegmHead, self).__init__()
+                     loss_weight=1.0),
+                 init_cfg=dict(type='Xavier', distribution='uniform')):
+        super(YOLACTSegmHead, self).__init__(init_cfg)
```
Comment (MeowZheng) [inline mmdet/models/dense_heads/yolact_head.py:476]: add layer

File: mmdet/models/dense_heads/yolact_head.py [hunk 10]
```diff
@@ -599,8 +591,9 @@ def __init__(self,
                  include_last_relu=True,
                  num_protos=32,
                  loss_mask_weight=1.0,
-                 max_masks_to_train=100):
-        super(YOLACTProtonet, self).__init__()
+                 max_masks_to_train=100,
+                 init_cfg=dict(type='Xavier', distribution='uniform')):
+        super(YOLACTProtonet, self).__init__(init_cfg)
```
Comment (MeowZheng) [inline mmdet/models/dense_heads/yolact_head.py:596]: add layer

File: mmdet/models/necks/hrfpn.py [hunk 11]
```diff
@@ -35,8 +36,15 @@ def __init__(self,
...
-        super(HRFPN, self).__init__()
+                 stride=1,
+                 init_cfg=dict(
+                     type='Kaiming',
+                     layer='Conv2d',
+                     a=1,
+                     mode='fan_in',
+                     nonlinearity='leaky_relu',
+                     distribution='uniform')):
+        super(HRFPN, self).__init__(init_cfg)
```
Comment (MeowZheng) [inline mmdet/models/necks/hrfpn.py:47]: long init_cfg

File: mmdet/models/necks/nas_fpn.py [hunk 12]
```diff
@@ -35,8 +36,15 @@ def __init__(self,
...
-        super(NASFPN, self).__init__()
+                 norm_cfg=None,
+                 init_cfg=dict(
+                     type='Kaiming',
+                     layer='Conv2d',
+                     a=1,
+                     mode='fan_in',
+                     nonlinearity='leaky_relu',
+                     distribution='uniform')):
+        super(NASFPN, self).__init__(init_cfg)
```
Comment (MeowZheng) [inline mmdet/models/necks/nas_fpn.py:47]: long init_cfg

File: mmdet/models/necks/rfp.py [hunk 13]
```diff
@@ -77,8 +75,21 @@ def __init__(self,
...
+                         distribution='uniform',
+                         override=[
+                             dict(name='lateral_convs', layer='Conv2d'),
+                             dict(name='fpn_convs', layer='Conv2d')
+                         ]),
+                     dict(
+                         type='Constant',
+                         val=0,
+                         override=dict(name='rfp_weight'))
+                 ],
```
Comment (MeowZheng) [inline mmdet/models/necks/rfp.py:90]: long init_cfg

File: mmdet/models/backbones/hourglass.py [hunk 14]
```diff
@@ -171,10 +176,13 @@ def init_weights(self, pretrained=None):
         Detector's __init__() will call backbone's init_weights() with
         pretrained as input, so we keep this function.
         """
-        # Training Centripetal Model needs to reset parameters for Conv2d
-        for m in self.modules():
-            if isinstance(m, nn.Conv2d):
-                m.reset_parameters()
+        if hasattr(self, 'init_cfg'):
```
Comment (jshilong) [inline mmdet/models/backbones/hourglass.py:179]: Please set `self._is_init` as True in all your overload.

**Comment by jshilong** [comment]
Please set `self._is_init` as True in all `init_weight`  your overload in this PR.

File: mmdet/models/necks/nasfcos_fpn.py [hunk 15]
```diff
@@ -146,16 +148,20 @@ def forward(self, inputs):
...
-        ]:
-            for module in modules:
-                if isinstance(module, nn.Conv2d):
-                    caffe2_xavier_init(module)
+        if hasattr(self, 'init_cfg'):
+            super(NASFCOS_FPN, self).init_weight()
+        else:
+            for module in self.fpn.values():
+                if hasattr(module, 'conv_out'):
+                    caffe2_xavier_init(module.out_conv.conv)
```
[the remark this answers is not in the document]
Comment (MeowZheng) [reply mmdet/models/necks/nasfcos_fpn.py:159]: mmcv [#902](https://github.com/open-mmlab/mmcv/pull/902)

File: mmdet/models/backbones/hourglass.py [hunk 16]
```diff
@@ -171,10 +176,13 @@ def init_weights(self, pretrained=None):
...
+            super(HourglassNet, self).init_weight()
+        else:
+            # Training Centripetal Model needs to reset parameters for Conv2d
+            for m in self.modules():
+                if isinstance(m, nn.Conv2d):
+                    m.reset_parameters()
```
[the remark this answers is not in the document]
Comment (MeowZheng) [reply mmdet/models/backbones/hourglass.py:185]: I strongly recommend override `init_weight()` with 'reset_parameters' and call `super()`

**Review by MeowZheng — COMMENTED** [review]
Many thanks for your nice work, it looks good.

File: mmdet/models/backbones/hrnet.py [hunk 17]
```diff
@@ -344,6 +350,34 @@ def __init__(self,
...
+                        layer=['_BatchNorm', 'GroupNorm'])
+                ]
+            elif not isinstance(init_cfg, list):
+                self.init_cfg = [init_cfg]
+
+            if self.zero_init_residual:
+                self.init_cfg += [
+                    dict(
+                        type='Constant',
+                        layer=['BatchNorm2', 'BatchNorm3'],
```
Comment (hhaAndroid) [inline mmdet/models/backbones/hrnet.py:375]: please confirm it

File: mmdet/models/backbones/detectors_resnet.py (hunk 2 above)
Comment (ZwwWayne) [reply mmdet/models/backbones/detectors_resnet.py:233]: We still need to keep back-compatibility for C4 models

File: mmdet/models/backbones/hrnet.py (hunk 17 above)
Comment (ZwwWayne) [reply mmdet/models/backbones/hrnet.py:375]: Is the name BatchNorm2 and BatchNorm3 correct?
The correctness can be verified by the model that training from scratch.

File: mmdet/models/backbones/regnet.py [hunk 18]
```diff
@@ -204,6 +211,36 @@ def __init__(self,
...
+                    dict(
+                        type='Constant',
+                        val=1,
+                        layer=['_BatchNorm', 'GroupNorm'])
+                ]
+            elif not isinstance(init_cfg, list):
+                self.init_cfg = [init_cfg]
+
+            # TODO: dcn conv_offset cannot initialize with init_cfg
+
```
Comment (ZwwWayne) [inline mmdet/models/backbones/regnet.py:233]: clean empty line.

File: mmdet/models/backbones/regnet.py [hunk 19]
```diff
@@ -204,6 +211,36 @@ def __init__(self,
...
+                    dict(type='Kaiming', layer='Conv2d'),
+                    dict(
+                        type='Constant',
+                        val=1,
+                        layer=['_BatchNorm', 'GroupNorm'])
+                ]
+            elif not isinstance(init_cfg, list):
+                self.init_cfg = [init_cfg]
+
+            # TODO: dcn conv_offset cannot initialize with init_cfg
```
Comment (ZwwWayne) [inline mmdet/models/backbones/regnet.py:232]: 1. dcn conv_offset cannot be initialized with init_cfg
2. cannot or should not?

**Comment by ZwwWayne** [comment]
1. The failed case of resnet zero initialization will be fixed in https://github.com/open-mmlab/mmcv/pull/918
2. The failed case of Sparse R-CNN is caused by the incorrect initialization of RPN proposals. Need to check further.

**Review by ZwwWayne — APPROVED** [review]

**Comment by shinya7y** [comment]
Will recent detectors be modified?
* Deformable DETR
* YOLOF
* AutoAssign
* Localization Distillation

**Comment by hhaAndroid** [comment]
> Will recent detectors be modified

No modification for the time being, if you think it is necessary, we can consider doing it in advance.

**Comment by shinya7y** [comment]
Thanks. My understanding is that modification is not essential (except for backbones).

**Comment by yeliudev** [comment]
@hhaAndroid May I ask that in the FPN module, all the conv layers are wrapped by `ConvModule`, which does NOT inherit from `BaseModule`. Will the `init_cfg` in FPN really affect the initialization behavior?
